### PR TITLE
add solution

### PR DESCRIPTION
**Pull request by SirWindfield**
I've added my changes to this branch. I attached some comments on what parts I would include from my solution into yours.

### CODE REVIEW

File: X-Sources/a1/Lechner/binpacking.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+def solve(items, bins, conflicts, capacity, weight):
+  # Model
+  model = Model("Binpacking with conflicts")
+
+  # Decision variable x_i_j indicates whether Item i is packed into Bin j (value 1) or not (value 0).
+  x = {}
+  for i in items:
+    for j in bins:
+      # TODO: Adjust additional attributes (lb, ub, vtype, obj). Do NOT change the name!
+      x[i,j] = model.addVar(name=f"x_{i}_{j}", vtype=GRB.BINARY)
```
Comment (SirWindfield) [inline X-Sources/a1/Lechner/binpacking.py]: Format string is imo a little bit nicer than calling `.format()`. It got introduced in Python 3.6 iirc.

File: X-Sources/a1/Lechner/binpacking.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  # Decision variable y_j indicates whether Bin j is used (value = 1) or not (value = 0).
+  y = {}
+  for j in bins:
+    # TODO: Adjust additional attributes (lb, ub, vtype, obj). Do NOT change the name!
+    y[j] = model.addVar(name=f"y_{j}", vtype=GRB.BINARY, obj=1.0)
```
Comment (SirWindfield) [inline X-Sources/a1/Lechner/binpacking.py]: Same as above

File: X-Sources/a1/Lechner/binpacking.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  # Update the model to make variables known. From now on, no variables should be added.
+  model.update()
+
+  # TODO: Add the linear constraints of the model. Nonlinearities in the model, e.g.,
+  # multiplication of two decision variables, results in a score of 0!
+
+  # Every item can only be packed once
+  # => sum over J starting at 1 with x_ij being equal to one for every item i in I
+  for i in range(len(items)):
+    is_in_bins = quicksum(x[i,j] for j in range(len(bins)))
```
Comment (SirWindfield) [inline X-Sources/a1/Lechner/binpacking.py]: I think appearances fits better as a name.

File: X-Sources/a1/Lechner/binpacking.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    model.addConstr(is_in_bins == 1)
+
+  # Each bin doesn't exceed its maximum capacity
+  # => sum over I starting at 1 with x_ij * weight_i being less or equal to c_j * y_j for every bin j in J
+  for j in range(len(bins)):
+    is_under_capacity = quicksum(x[i,j] * weight[i] for i in range(len(items)))
```
Comment (SirWindfield) [inline X-Sources/a1/Lechner/binpacking.py]: I don't like either variable name. I'm just being finicky right now :p

File: X-Sources/a1/Lechner/binpacking.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    model.addConstr(is_under_capacity <= capacity * y[j]) # looks like each bin has the same capacity...
+
+  # Some items can't be packed with eachother. Constraints are given as tuples
+  # => for each bin, check if there is only one of the items in them (valid values are 0 and 1)
+  for idx_item1, idx_item2 in conflicts:
+    for j in range(len(bins)):
+      model.addConstr(x[idx_item1,j] + x[idx_item2,j] <= 1)
```
Comment (SirWindfield) [inline X-Sources/a1/Lechner/binpacking.py]: Not gonna lie, this one took me a while haha :D

File: X-Sources/a1/Lechner/binpacking.py (hunk 1 above)
Comment (RCoanda) [reply X-Sources/a1/Lechner/binpacking.py]: I agree, still used to the old `.format()` style.

**Pull request merged**